sql: timestamp should accept DB2 formatted timestamps

sooryranga opened the pull request:
fixes : #49768

This PR adds additional extracting field logics that
allow db2 formatted timesstamps to be parsed.

Release note (sql change): The pgdate util can now parse
db2 formatted timestamps. This can be used to import db2 dumps.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=50011) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=50011) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=50011) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50011)
<!-- Reviewable:end -->

irfansharif commented:
@jordanlewis: feel free to re-assign, not sure who's the expert here now.

otan reviewed (commented):
Thanks for contributing! I've got a few questions with regards to how it affects existing cases, as well as a request for more tests.

Can you also add a test in https://github.com/cockroachdb/cockroach/blob/d9ae34b2851843ec2c350436f9cf7add550b0273/pkg/sql/sem/tree/testdata/eval/cast#L490? `go test ./pkg/sql/sem/tree -run 'TestEval/cast'` should trigger it.

otan commented on pkg/util/timeutil/pgdate/parsing_test.go:
```diff
@@ -632,6 +632,10 @@ var timestampTestData = []timeData{
 		hasTimezone: true,
 	},
+	{
+		s:   "2014-03-15-04.38.53.399853+07:35",
```
a few error conditions, e.g. `2007-09-24-15.0` may also be nice to test.

Arun4rangan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -252,6 +252,7 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 		//                   ^^^^^^
 		// Example: 1999.123, a year + day-of-year.
 		//               ^^^
+		// Example: 04.04.04.913231+00:00, odb2 timestamp
```
Good catch!

Arun4rangan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -451,7 +459,10 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 		return fe.SetChunk(fieldYear, chunk)
-
+	case !fe.Wants(fieldDay) && fe.Wants(fieldHour) && chunk.separator == '-':
+		// Example: "YYYY-MM-DD-HH-MM-SS"
```
No there isn't, it's a mistake.

otan commented on pkg/util/timeutil/pgdate/parsing.go:
```diff
@@ -55,6 +55,8 @@ var (
 		fieldTZHour, fieldTZMinute, fieldTZSecond)
 	timeRequiredFields = newFieldSet(fieldHour, fieldMinute)
 
+	odbTimeRequiredFields = newFieldSet(fieldHour, fieldMinute, fieldSecond)
```
any reason this is called `odb`?

otan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -246,12 +248,17 @@ func (fe *fieldExtract) Get(field field) (int, bool) {
 		//                   ^^^^^^
 		// Example: 1999.123, a year + day-of-year.
 		//               ^^^
+		// Example: 04.04.04.913231+00:00, odb timestamp
```
what is `odb`?

otan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -263,6 +270,14 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
+		case fe.Wants(fieldMinute) && chunk.v <= 60 && chunk.v >= 0:
+			// db2 timestamp allows for minutes to recorded with periods.
+			// Example 13.12.50 (hh.mm.ss)
+			fe.isDB2 = true
```
should `fe.isDB2` already be set by the `-` separator below, or is the `-` between the YYYY-MM-DD and HH.MM.SS optional?

Arun4rangan replied:
You could have something like `yyyy-mm-dd hh.mm.ss` which is a DB2 type, we need to set DB2 on `mm` so if we get something like `yyyy-mm-dd hh.mm:ss` we could throw an error. 

In this cause it won't `-` but still should be treated as db2 timestamp.

otan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -64,6 +64,8 @@ type fieldExtract struct {
 	tzSign int
 	// Tracks the fields that we want to extract.
 	wanted fieldSet
+	// Tracks whether the current timestamp formes db2 format
```
nit: formes seems out of places. "Tracks whether the current timestamp is of db2 format.", perhaps.

otan commented on pkg/storage/enginepb/mvcc3.pb.go:
```diff
@@ -4207,7 +4207,9 @@ var (
 	ErrIntOverflowMvcc3   = fmt.Errorf("proto: integer overflow")
 )
 
-func init() { proto.RegisterFile("storage/enginepb/mvcc3.proto", fileDescriptor_mvcc3_9bc532bf92053320) }
```
please revert these protobuf .pb.go changes

otan commented on pkg/sql/sem/tree/testdata/eval/cast:
```diff
@@ -1070,3 +1070,48 @@ eval
+eval
+time '1969-12-30-01.00.00.12345'
+----
+'01:00:00.12345'
```
please add \n at end of file.

otan commented on pkg/sql/sem/tree/testdata/eval/cast:
```diff
@@ -1070,3 +1070,48 @@ eval
 -10.000000000
+
+eval
+'1969-12-30-01.00.00'::timestamp
```
please put these near the other timezone cases. they do not have to be exhaustive either -- just one or two is fine (these should be unit tests instead if we need them)

otan reviewed (commented):
Thanks! couple more small nits.

sooryranga committed:
sql: timestamp should accept DB2 formatted timestamps
fixes : #49768

This PR adds additional extracting field logics that
allow db2 formatted timesstamps to be parsed.

Release note (sql change): The pgdate util can now parse
db2 formatted timestamps. This can be used to import db2 dumps.

Release note (<category, see below>): <what> <show> <why>

Arun4rangan commented on pkg/util/timeutil/pgdate/field_extract.go:
```diff
@@ -263,6 +264,13 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 			// BUT NOT: 1999 1
 			return fe.SetDayOfYear(chunk)
 
+		case fe.Wants(fieldMinute) && chunk.v <= 60 && chunk.v >= 0:
```
You are right, using `isDB2` as a trigger prevents unnecessary field extraction.

otan reviewed (approved):
Thanks!
